src/coreclr/jit/importer_vectorization.cpp

```diff
@@ -0,0 +1,741 @@
+
+    GenTree* op1;
+    GenTree* op2;
+    if (argsCount == 3) // overload with StringComparison
+    {
+        if (!impStackTop(0).val->IsIntegralConst(4)) // StringComparison.Ordinal
+        {
+            // TODO-Unroll-CQ: Unroll & vectorize OrdinalIgnoreCase
+            return nullptr;
+        }
+        op1 = impStackTop(2).val;
+        op2 = impStackTop(1).val;
+    }
+    else
+    {
+        assert(argsCount == 2);
+        op1 = impStackTop(1).val;
+        op2 = impStackTop(0).val;
+    }
+
+    if (!(op1->OperIs(GT_CNS_STR) ^ op2->OperIs(GT_CNS_STR)))
+    {
+        // either op1 or op2 has to be CNS_STR, but not both - that case is optimized
+        // just fine as is.
+        return nullptr;
+    }
+
+    GenTree*       varStr;
+    GenTreeStrCon* cnsStr;
+    if (op1->OperIs(GT_CNS_STR))
+    {
+        cnsStr = op1->AsStrCon();
+        varStr = op2;
+    }
+    else
+    {
+        cnsStr = op2->AsStrCon();
+        varStr = op1;
+    }
+
+    bool needsNullcheck = true;
+    if ((op1 != cnsStr) && !isStatic)
+    {
+        // for the following cases we should not check varStr for null:
+        //
+        //  obj.Equals("cns")
+        //  obj.Equals("cns", StringComparison.Ordinal)
+        //  obj.StartsWith("cns", StringComparison.Ordinal)
+        //
+        // instead, it should throw NRE if it's null
+        needsNullcheck = false;
+    }
+
+    int             cnsLength = -1;
+    const char16_t* str       = nullptr;
+    if (cnsStr->IsStringEmptyField())
+    {
+        // check for fake "" first
+        cnsLength = 0;
+        JITDUMP("Trying to unroll String.Equals|StartsWith(op1, \"\")...\n", str)
+    }
+    else
+    {
+        str = info.compCompHnd->getStringLiteral(cnsStr->gtScpHnd, cnsStr->gtSconCPX, &cnsLength);
+        if (cnsLength < 0 || str == nullptr)
+        {
+            // We were unable to get the literal (e.g. dynamic context)
+            return nullptr;
+        }
+        JITDUMP("Trying to unroll String.Equals|StartsWith(op1, \"%ws\")...\n", str)
+    }
+
+    // Create a temp which is safe to gtClone for varStr
+    // We're not appending it as a statement until we figure out unrolling is profitable (and possible)
+    unsigned varStrTmp         = lvaGrabTemp(true DEBUGARG("spilling varStr"));
+    lvaTable[varStrTmp].lvType = varStr->TypeGet();
+    GenTree* varStrLcl         = gtNewLclvNode(varStrTmp, varStr->TypeGet());
+
+    // Create a tree representing string's Length:
+    // TODO-Unroll-CQ: Consider using ARR_LENGTH here, but we'll have to modify QMARK to propagate BBF_HAS_IDX_LEN
+    int      strLenOffset = OFFSETOF__CORINFO_String__stringLen;
+    GenTree* lenOffset    = gtNewIconNode(strLenOffset, TYP_I_IMPL);
+    GenTree* lenNode      = gtNewIndir(TYP_INT, gtNewOperNode(GT_ADD, TYP_BYREF, varStrLcl, lenOffset));
+
+    GenTree* unrolled = impExpandHalfConstEquals(gtClone(varStrLcl), lenNode, needsNullcheck, startsWith, (WCHAR*)str,
```

Comment:
If you are relying on this to avoid the issues noted above with `gtClone` then you should assert in those methods that you have GT_LCL_VARs and not some other tree.

Reply:
Changed data argument to be LclVar in the signature so it's cheap to clone by default now